drivers/wifi/esp8266/esp8266.c

```diff
@@ -0,0 +1,784 @@
+		return -EALREADY;
+	}
+
+	esp8266_flag_set(data, EDF_STA_CONNECTING);
+
+	len = snprintk(cmd, sizeof(cmd), "AT+CWJAP_CUR=\"");
+	memcpy(&cmd[len], params->ssid, params->ssid_length);
+	len += params->ssid_length;
+
+	if (params->security == WIFI_SECURITY_TYPE_PSK) {
+		len += snprintk(&cmd[len], sizeof(cmd) - len, "\",\"");
+		memcpy(&cmd[len], params->psk, params->psk_length);
+		len += params->psk_length;
+	}
+
+	len += snprintk(&cmd[len], sizeof(cmd) - len, "\"");
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, cmd, &data->sem_response,
+			     K_NO_WAIT);
+
+	esp8266_flag_clear(data, EDF_STA_CONNECTING);
+
+	return ret;
+}
+
+static int esp8266_mgmt_disconnect(struct device *dev)
+{
+	struct esp8266_data *data = dev->driver_data;
+	int ret;
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, "AT+CWQAP", &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static int esp8266_mgmt_ap_enable(struct device *dev,
+				  struct wifi_connect_req_params *params)
+{
+	char cmd[sizeof("AT+CWSAP_CUR=\"\",\"\",xx,x") + WIFI_SSID_MAX_LEN +
+		 WIFI_PSK_MAX_LEN];
+	struct esp8266_data *data = dev->driver_data;
+	int ecn = 0, len, ret;
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, "AT+CWMODE_CUR=3", &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("Failed to enable AP mode, ret %d", ret);
+		return ret;
+	}
+
+	len = snprintk(cmd, sizeof(cmd), "AT+CWSAP_CUR=\"");
+	memcpy(&cmd[len], params->ssid, params->ssid_length);
+	len += params->ssid_length;
+
+	if (params->security == WIFI_SECURITY_TYPE_PSK) {
+		len += snprintk(&cmd[len], sizeof(cmd) - len, "\",\"");
+		memcpy(&cmd[len], params->psk, params->psk_length);
+		len += params->psk_length;
+		ecn = 3;
+	} else {
+		len += snprintk(&cmd[len], sizeof(cmd) - len, "\",\"");
+	}
+
+	snprintk(&cmd[len], sizeof(cmd) - len, "\",%d,%d", params->channel,
+		 ecn);
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, cmd, &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static int esp8266_mgmt_ap_disable(struct device *dev)
+{
+	struct esp8266_data *data = dev->driver_data;
+	int ret;
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, "AT+CWMODE_CUR=1", &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static void esp8266_init_work(struct k_work *work)
+{
+	struct esp8266_data *dev;
+	int ret;
+	static struct setup_cmd setup_cmds[] = {
+		/* turn off echo */
+		SETUP_CMD_NOHANDLE("ATE0"),
+		/* enable multiple socket support */
+		SETUP_CMD_NOHANDLE("AT+CIPMUX=1"),
+		SETUP_CMD_NOHANDLE("AT+CWMODE_CUR=1"),
+		/* only need ecn,ssid,rssi,channel */
+		SETUP_CMD_NOHANDLE("AT+CWLAPOPT=0,23"),
+#if defined(CONFIG_WIFI_ESP8266_PASSIVE_TCP)
+		SETUP_CMD_NOHANDLE("AT+CIPRECVMODE=1"),
+#endif
+		SETUP_CMD("AT+CIPAPMAC_CUR?", "+CIPAPMAC_CUR:",
```

Comment:
Should this be **CIPSTA** as this is a station or do we need to consider logic to define AT command if Access Point (AP) or Client (Station)?

Reply:
Yea sure lets use station mac for now. Not sure how to handle the link address when AP is started since you can be both STA and AP.